src/Aspire.Hosting/Health/ResourceHealthCheckService.cs

```diff
@@ -0,0 +1,119 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Immutable;
+using Aspire.Hosting.ApplicationModel;
+using Microsoft.Extensions.Diagnostics.HealthChecks;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+
+namespace Aspire.Hosting.Health;
+
+internal class ResourceHealthCheckService(ILogger<ResourceHealthCheckService> logger, ResourceNotificationService resourceNotificationService, HealthCheckService healthCheckService) : BackgroundService
+{
+    private readonly Dictionary<string, Task> _resourceTasks = new();
+    private readonly Dictionary<string, ResourceEvent> _latestEvents = new();
```

Comment:
I don't love these being fields if they aren't used outside of these 2 methods.

Reply:
We need them. I've changed the resource tasks one to a hashset since I don't (currently) need access to the tasks. But I do need the latest snapshot for each resource to determine whether I should publish a status update (to avoid noise).

Reply:
Make sure there’s no concurrency issues

Reply:
Just realized I could move one into a local variable due to some refactoring I did.

The remaining one is only updated from the main loop, the monitoring loops just read from it periodically so its safe. We don't ever remove resources so we don't need to worry about an event not being present by the time the monitoring loop for a resource is up and running.